# Linear programming and discrete optimization using Python with PuLP

In [1]:
# !pip install pulp

In [2]:
from pulp import *
import pandas as pd

C:\Users\User\anaconda3\lib\site-packages\numpy\_distributor_init.py:32: UserWarning: loaded more than 1 DLL from .libs:
C:\Users\User\anaconda3\lib\site-packages\numpy\.libs\libopenblas.NOIJJG62EMASZI6NYURL6JBKM4EVBGM7.gfortran-win_amd64.dll
C:\Users\User\anaconda3\lib\site-packages\numpy\.libs\libopenblas.PYQHXLVVQ7VESDPUVUADXEVJOBGHJPAY.gfortran-win_amd64.dll
  stacklevel=1)


### Load and Prepare Data

In [3]:
# Read the first few rows dataset in a Pandas DataFrame
# Read only the nutrition info not the bounds/constraints
df = pd.read_excel("Food.xlsx",nrows=17)

In [4]:
df.head(5)

Foods  Price/Serving  Calories  Total_Fat(g)  Protein  \
0      Forzen Broccoli           0.48      73.8           0.8      8.0   
1          Forzen Corn           0.54      72.2           0.6      2.5   
2  Raw Lattuce Iceberg           0.06       2.6           0.0      0.2   
3       Baked Potatoes           0.18     171.5           0.2      3.7   
4                 Tofu           0.93      88.2           5.5      9.4   

   Carbohydrates(g)  Dietary_Fiber(g)  
0              13.6               8.5  
1              17.1               2.0  
2               0.4               0.3  
3              39.9               3.2  
4               2.2               1.4

In [5]:
# Create a list of the food items
food_items = list(df['Foods'])

In [6]:
food_items[:5]

['Forzen Broccoli',
 'Forzen Corn',
 'Raw Lattuce Iceberg',
 'Baked Potatoes',
 'Tofu']

In [7]:
# Create a dictinary of costs for all food items
costs = dict(zip(food_items,df['Price/Serving']))

In [8]:
costs

{'Forzen Broccoli': 0.48,
 'Forzen Corn': 0.54,
 'Raw Lattuce Iceberg': 0.06,
 'Baked Potatoes': 0.18,
 'Tofu': 0.93,
 'Roasted Chicken': 2.52,
 'Spaghetti W/Sauce': 2.34,
 'Raw Apple': 0.72,
 'Banana': 0.45,
 'Wheat Bread': 0.15,
 'White Bread': 0.18,
 'Oatmeal Cookies': 0.27,
 'Apple Pie': 0.48,
 'Scrambled Eggs': 0.33,
 'Turkey Bologna': 0.45,
 'Beef Frankfurter': 0.81,
 'Chocolates Chip Cookies': 0.09}

In [9]:
# Create a dictionary of calories for all food items
calories = dict(zip(food_items,df['Calories']))

In [10]:
# Create a dictionary of total fat for all food items
fat = dict(zip(food_items,df['Total_Fat(g)']))

In [11]:
# Create a dictionary of preteins for all food items
protein = dict(zip(food_items,df['Protein']))

In [12]:
# Create a dictionary of carbohydrates for all food items
carbs = dict(zip(food_items,df['Carbohydrates(g)']))

In [13]:
# Create a dictionary of fibres for all food items
fibers = dict(zip(food_items,df['Dietary_Fiber(g)']))

### Formulating the problem

In [14]:
# First, we create a LP problem with the method LpProblemin PuLP.
prob = LpProblem("Simple Diet Problem", LpMinimize)

C:\Users\User\anaconda3\lib\site-packages\pulp\pulp.py:1198: UserWarning: Spaces are not permitted in the name. Converted to '_'
  warnings.warn("Spaces are not permitted in the name. Converted to '_'")


In [15]:
prob

Simple_Diet_Problem:
MINIMIZE
None
VARIABLES

In [16]:
'''
Then, we create a dictionary of food items variables with lower bound =0 and category continuous 
i.e. the optimization solution can take any real-numbered value greater than zero.

The particular importance of the lower bound is that in practice food quantities cannot be negative
'''

food_vars = LpVariable.dicts("Food", food_items, lowBound=0, cat='Continuous')

In [17]:
food_vars

{'Forzen Broccoli': Food_Forzen_Broccoli,
 'Forzen Corn': Food_Forzen_Corn,
 'Raw Lattuce Iceberg': Food_Raw_Lattuce_Iceberg,
 'Baked Potatoes': Food_Baked_Potatoes,
 'Tofu': Food_Tofu,
 'Roasted Chicken': Food_Roasted_Chicken,
 'Spaghetti W/Sauce': Food_Spaghetti_W_Sauce,
 'Raw Apple': Food_Raw_Apple,
 'Banana': Food_Banana,
 'Wheat Bread': Food_Wheat_Bread,
 'White Bread': Food_White_Bread,
 'Oatmeal Cookies': Food_Oatmeal_Cookies,
 'Apple Pie': Food_Apple_Pie,
 'Scrambled Eggs': Food_Scrambled_Eggs,
 'Turkey Bologna': Food_Turkey_Bologna,
 'Beef Frankfurter': Food_Beef_Frankfurter,
 'Chocolates Chip Cookies': Food_Chocolates_Chip_Cookies}

In [18]:
# Next, we start building the LP problem by adding the main objective function. Note the use of the lpSum method.
prob += lpSum([costs[i]*food_vars[i] for i in food_items])

In [19]:
prob

Simple_Diet_Problem:
MINIMIZE
0.48*Food_Apple_Pie + 0.18*Food_Baked_Potatoes + 0.45*Food_Banana + 0.81*Food_Beef_Frankfurter + 0.09*Food_Chocolates_Chip_Cookies + 0.48*Food_Forzen_Broccoli + 0.54*Food_Forzen_Corn + 0.27*Food_Oatmeal_Cookies + 0.72*Food_Raw_Apple + 0.06*Food_Raw_Lattuce_Iceberg + 2.52*Food_Roasted_Chicken + 0.33*Food_Scrambled_Eggs + 2.34*Food_Spaghetti_W_Sauce + 0.93*Food_Tofu + 0.45*Food_Turkey_Bologna + 0.15*Food_Wheat_Bread + 0.18*Food_White_Bread + 0.0
VARIABLES
Food_Apple_Pie Continuous
Food_Baked_Potatoes Continuous
Food_Banana Continuous
Food_Beef_Frankfurter Continuous
Food_Chocolates_Chip_Cookies Continuous
Food_Forzen_Broccoli Continuous
Food_Forzen_Corn Continuous
Food_Oatmeal_Cookies Continuous
Food_Raw_Apple Continuous
Food_Raw_Lattuce_Iceberg Continuous
Food_Roasted_Chicken Continuous
Food_Scrambled_Eggs Continuous
Food_Spaghetti_W_Sauce Continuous
Food_Tofu Continuous
Food_Turkey_Bologna Continuous
Food_Wheat_Bread Continuous
Food_White_Bread Continuous

In [20]:
# We further build on this by adding calories constraints
prob += lpSum([calories[f] * food_vars[f] for f in food_items]) >= 800.0
prob += lpSum([calories[f] * food_vars[f] for f in food_items]) <= 1300.0

In [21]:
prob

Simple_Diet_Problem:
MINIMIZE
0.48*Food_Apple_Pie + 0.18*Food_Baked_Potatoes + 0.45*Food_Banana + 0.81*Food_Beef_Frankfurter + 0.09*Food_Chocolates_Chip_Cookies + 0.48*Food_Forzen_Broccoli + 0.54*Food_Forzen_Corn + 0.27*Food_Oatmeal_Cookies + 0.72*Food_Raw_Apple + 0.06*Food_Raw_Lattuce_Iceberg + 2.52*Food_Roasted_Chicken + 0.33*Food_Scrambled_Eggs + 2.34*Food_Spaghetti_W_Sauce + 0.93*Food_Tofu + 0.45*Food_Turkey_Bologna + 0.15*Food_Wheat_Bread + 0.18*Food_White_Bread + 0.0
SUBJECT TO
_C1: 67.2 Food_Apple_Pie + 171.5 Food_Baked_Potatoes + 104.9 Food_Banana
 + 141.8 Food_Beef_Frankfurter + 78.1 Food_Chocolates_Chip_Cookies
 + 73.8 Food_Forzen_Broccoli + 72.2 Food_Forzen_Corn + 81 Food_Oatmeal_Cookies
 + 81.4 Food_Raw_Apple + 2.6 Food_Raw_Lattuce_Iceberg
 + 277.4 Food_Roasted_Chicken + 99.6 Food_Scrambled_Eggs
 + 358.2 Food_Spaghetti_W_Sauce + 88.2 Food_Tofu + 56.4 Food_Turkey_Bologna
 + 65 Food_Wheat_Bread + 65 Food_White_Bread >= 800

_C2: 67.2 Food_Apple_Pie + 171.5 Food_Baked_Potatoes

In [22]:
# We can pile up all the nutrition constraints.

# Fat
prob += lpSum([fat[f] * food_vars[f] for f in food_items]) >= 20.0, "FatMinimum"
prob += lpSum([fat[f] * food_vars[f] for f in food_items]) <= 50.0, "FatMaximum"

In [23]:
# Carbs
prob += lpSum([carbs[f] * food_vars[f] for f in food_items]) >= 130.0, "CarbsMinimum"
prob += lpSum([carbs[f] * food_vars[f] for f in food_items]) <= 200.0, "CarbsMaximum"

In [24]:
# Fiber
prob += lpSum([fibers[f] * food_vars[f] for f in food_items]) >= 60.0, "FiberMinimum"
prob += lpSum([fibers[f] * food_vars[f] for f in food_items]) <= 125.0, "FiberMaximum"

In [25]:
# Protein
prob += lpSum([protein[f] * food_vars[f] for f in food_items]) >= 100.0, "ProteinMinimum"
prob += lpSum([protein[f] * food_vars[f] for f in food_items]) <= 150.0, "ProteinMaximum"

At this point, formulating the problem is done.

##### In any optimization scenario, the hard part is the formulation of the problem in a structured manner which is presentable to a solver.

### Solving the problem

In [26]:
'''
PuLP has quite a few choices of solver algorithms (e.g. COIN_MP, Gurobi, CPLEX, etc.). 
For this problem, we do not specify any choice and let the program default to its own choice 
depending on the problem structure.
'''

prob.solve() # 

1

In [27]:
# The status of the solution is printed to the screen
print("Status:", LpStatus[prob.status])

Status: Optimal


In [28]:
'''
The full solution contains all the variables including the ones with zero weights. 
But to us, only those variables are interesting which have non-zero coefficients 
i.e. which should be included in the optimal diet plan. 
So, we can scan through the problem variables and print out only if the variable quantity is positive.
'''

for v in prob.variables():
    if v.varValue>0:
        print(v.name, "=", v.varValue)

Food_Baked_Potatoes = 1.0806324
Food_Forzen_Broccoli = 6.9242113
Food_Scrambled_Eggs = 6.060891


So, the optimal solution is to eat 6.924 servings of frozen broccoli, 6.06 servings of scrambled eggs and 1.08 servings of a baked potato!

Reference https://towardsdatascience.com/linear-programming-and-discrete-optimization-with-python-using-pulp-449f3c5f6e99